### ISSUE TITLE
Localising option on (un)setting automatic introduction

### ISSUE THREAD

**Issue by benediktahrens** [UniMath/UniMath#323]
The option "Unset Automatic Introduction" was introduced in Coq 8.3 to create backwards compat with 8.2.
Used without a "Local" modifier, this option modifies behaviour of Coq upon Requiring a module changing that setting.
It seems to me and others that this setting should always be used with a "Local" modifier in order to obtain a predictible behaviour - namely, the default, which is Automatic Introduction - when not explicitly changing this setting.

In short: all the uses of this option should be Local to the file/module, imo.

Thoughts?

**Comment by mortberg** [comment]
I agree that it should be Local. Or if noone is really relying on this option we could just remove it, at least to me the standard behaviour is the most intuitive. Also for someone trying to learn UniMath the fewer strange commands the better.

**Comment by benediktahrens** [comment]
On 03/18/2016 08:02 AM, Anders Mörtberg wrote:
> I agree that it should be Local. Or if noone is really relying on
> this option we could just remove it, at least to me the standard
> behaviour is the most intuitive. Also for someone trying to learn
> UniMath the fewer strange commands the better.

I suppose that you will have to change many files in many places if you
want to get rid of that option totally.

Also, an author of a file should still be able to use the settings they
prefer, so I don't see any reason to try to get rid of that command
entirely.

The restriction to use it only with "Local" seems both reasonable and
sufficient to me.

**Comment by mortberg** [comment]
I was just thinking that if it was added to create backwards compatibility with 8.2 it might not be worth keeping it. But if some files are really relying on it making it Local to those files seems like the best option.

**Comment by DanGrayson** [comment]
Automatic introduction is a good thing, so having it as the default would be good, and we might want to just convert all the files to use it.

**Comment by vladimirias** [comment]
I rely on it. However, I agree that it should be local.

**Comment by vladimirias** [comment]
I think I disagree. Writing one “intro.” is really easy while dealing with the cases where one for induction needs partial introduction when there is automated full introduction is bothersome.

**Comment by benediktahrens** [comment]
I would like to advertise this issue. We should have consistent behaviour throughout, which should only be changed on a file-by-file basis.
I was bitten by this in https://github.com/UniMath/UniMath/pull/673 , where I removed an unnecessary Import in one file, and that changed the introduction of variables in another file importing the changed file. This is unexpected behaviour and should not happen.

**Comment by benediktahrens** [comment]
Specifically, what happened is that I removed a `Require Import UniMath.Ktheory.Utilities.` in Topology.Prelim, which made Topology.Topology fail to compile (but not Topology.Prelim). That's silly.

**Comment by DanGrayson** [comment]
Vladimir, you say "I agree that it should be local."

Do you mean we can change the 9 files in Foundations that contain the command "Unset Automatic Introduction" so they are all local?

**Comment by benediktahrens** [comment]
I would like to revive this issue.
I suggest:
- Fixing a default, set in Preamble (if necessary)
- Any calls changing this option must be prefixed by `Local`.
I have a slight preference to set the default to `Set Automatic Introduction`, but more important than that is consistency.

**Comment by DanGrayson** [comment]
I agree.  This affects mostly the Foundation directory, so I'll do it that.  Assigning myself.

**Comment by benediktahrens** [comment]
We seem to have zero uses of "(Un)Set Automatic Introduction", so this is not an issue any more - closing.
